CMakeLists.txt

```diff
@@ -622,22 +631,24 @@ if(WITH_OPTIM)
         add_definitions(-DARM_GETAUXVAL)
         list(APPEND ZLIB_ARCH_SRCS ${ARCHDIR}/armfeature.c ${ARCHDIR}/fill_window_arm.c)
         if(WITH_NEON)
-            list(APPEND ZLIB_ARCH_SRCS ${ARCHDIR}/adler32_neon.c)
             add_definitions(-DARM_NEON_ADLER32)
-            add_intrinsics_option("${NEONFLAG}")
             if(MSVC)
                 add_definitions(-D__ARM_NEON__)
             endif()
+            set(NEON_SRCS ${ARCHDIR}/adler32_neon.c)
+            list(APPEND ZLIB_ARCH_SRCS ${NEON_SRCS})
+            add_intrinsics_option(${NEONFLAG} ${NEON_SRCS})
             add_feature_info(NEON_FILLWINDOW 1 "Support NEON instructions in fill_window_arm, using \"${NEONFLAG}\"")
         endif()
         if(WITH_ACLE)
-            list(APPEND ZLIB_ARCH_SRCS ${ARCHDIR}/crc32_acle.c ${ARCHDIR}/insert_string_acle.c)
             add_definitions(-DARM_ACLE_CRC_HASH)
+            set(ACLE_SRCS ${ARCHDIR}/crc32_acle.c ${ARCHDIR}/insert_string_acle.c)
+            list(APPEND ZLIB_ARCH_SRCS ${ACLE_SRCS})
             # For ARM aarch64, we need to check WITH_NEON first
-            if("${ARCH}" MATCHES "arm" OR NOT WITH_NEON)
-                add_intrinsics_option("${ACLEFLAG}")
+            if(${ARCH} MATCHES "arm" OR NOT WITH_NEON)
```

Comment:
Removing quotes might be unsafe if the variable contents have non-alphanumeric characters which is possible under ARM. Basically variable expansion happens before IF condition is evaluated and it can change the order the conditions are executed.

Reply:
Fixed